clang/lib/Sema/SemaDecl.cpp

```diff
@@ -9371,8 +9371,11 @@ Sema::ActOnFunctionDeclarator(Scope *S, Declarator &D, DeclContext *DC,
         // (void) parameters, so we relax this to a warning.
         int DiagID =
             CC == CC_X86StdCall ? diag::warn_cconv_knr : diag::err_cconv_knr;
-        Diag(NewFD->getLocation(), DiagID)
```

Comment:
I think this needs to be in an 'else'. Otherwise wouldn't behavior change for non-SYCL code?

Reply:
- Invoking `dpcpp-cl test.c`, causes the compiler to compile `test.c` as host and device code, even though it is just a C source file (non-sycl code)
- no-prototype functions are allowed in C. While it is not a good coding style, it should not be causing compilation failure.
- `Diag(NewFD->getLocation(), DiagID)` causes compilation failure.
- Because dpcpp-cl invokes clang-cl with -cc1, -fsycl, -fsycl-is-device and a bunch of other options, getLangOpts().SYCLIsDevice is true. 
- "else" clause if present, will never be entered. 
- This block of code is inside `if(!getLangOpts().CPlusPlus)` - meaning for C and Obj-C languages?
- And if np-functions are allowed in C (and possible Obj-C), I am not sure if we should emit the Diag.

Reply:
@premanandrao 

- While debugging this change against a SYCL application ( C++ source code file) whose kernel invokes a no-prototype function, because of this check
 `if(!getLangOpts().CPlusPlus)` , we never enter this block. 
- And this never gets executed - `SYCLDiagIfDeviceCode(D.getIdentifierLoc(), DiagID) << FunctionType::getNameForCallConv(CC);`
I am not sure if we should be checking for device code calling np-functions here.

Reply:
Removing this diagnostic entirely is not right.

Reply:
This seems to contradict what we discussed offline.
Per our discussion - The use of function declarators with empty parentheses is allowed in C source code but not inside a SYCL kernel. 
Based on the above statement, changes were made to detect if no-prototype functions were used in SYCL kernel and if they were used, a deferred diagnostic is emitted.
The original Diag was removed since C does not restrict no-prototype functions.

The problem with the above change is - the check for np-function in SYCL kernel is added inside a `if(!getLangOpts().CPlusPlus)` which will never get executed since SYCL apps are C++ programs.

Few clarifications that came from Mariya's comment are:
SYCL kernel does not restrict the usage of function declarators with empty parentheses.
This makes the emission of deferred diagnostic unnecessary.